src/mozilla_sec_eia/models/sec_eia_record_linkage/transform_sec_input.py

```diff
@@ -0,0 +1,371 @@
+    sec_df.loc[:, "report_year"] = (
+        sec_df["report_date"].astype("datetime64[ns]").dt.year
+    )
+    return sec_df
+
+
+def _flatten_sec_companies_across_time(sec_df: pd.DataFrame) -> pd.DataFrame:
+    """Keep only the most recent record for each unique SEC company.
+
+    Note that this drops old records for companies that have changed
+    names or addresses across time. Also, we group by sec_company_id not
+    CIK, so filer companies and subsidiary companies are unique in the
+    output dataframe.
+    TODO: create an asset that tracks name and address chnages across
+    time.
+    """
+    sec_df = (
+        sec_df.sort_values(by="report_year", ascending=False)
+        .groupby("sec_company_id")
+        .first()
+    ).reset_index()
+    return sec_df
+
+
+def get_sec_state_code_dict() -> dict[str, str]:
+    """Create a dictionary mapping state codes to their names.
+
+    Table found at https://www.sec.gov/submit-filings/filer-support-resources/edgar-state-country-codes
+    Published by SEC and reports valid state codes
+    for filers of Form D. Used to standardize the state codes
+    in the SEC 10K filings. The expanded names of the state codes
+    are comments in the XML file, so we have to read the XML in as
+    text and parse it.
+    """
+    # TODO: make a check to see if SEC has published a new version of this table
+    xml_filepath = (
+        resources.files("mozilla_sec_eia.package_data") / "formDStateCodes.xsd.xml"
+    )
+    with Path.open(xml_filepath) as file:
+        xml_text = file.read()
+
+    pattern = r'<xs:enumeration value="(.*?)"/>.*?<!--\s*(.*?)\s*-->'
+    state_code_dict = {
+        code.lower(): name.lower()
+        for code, name in re.findall(pattern, xml_text, re.DOTALL)
+    }
+    return state_code_dict
+
+
+def clean_location_of_inc(df) -> pd.DataFrame:
+    """Clean location of incorporation column in SEC basic 10K or Ex. 21 dataframe.
+
+    Arguments:
+        df: Ex. 21 or SEC 10K basic info dataframe with location_of_inc
+            column.
+    """
+    if "state_of_incorporation" in df:
+        df.loc[:, "location_of_inc"] = df["state_of_incorporation"]
+    state_code_to_name = get_sec_state_code_dict()
+    df.loc[:, "location_of_inc"] = (
+        df["location_of_inc"]
+        .replace(state_code_to_name)
+        .fillna(pd.NA)
+        .str.strip()
+        .str.lower()
+        .replace("", pd.NA)
+    )
+    return df
+
+
+def _add_parent_company_cik(ex21_df: pd.DataFrame, md: pd.DataFrame) -> pd.DataFrame:
+    """Add the CIK of the parent company to Ex. 21 subsidiaries."""
+    ex21_df = ex21_df.merge(md[["filename", "cik"]], how="left", on="filename").rename(
+        columns={"cik": "parent_company_cik"}
+    )
+    ex21_df.loc[:, "parent_company_cik"] = (
+        ex21_df["parent_company_cik"].astype(str).str.zfill(10)
+    )
+    return ex21_df
+
+
+def match_ex21_subsidiaries_to_filer_company(
```

Comment:
This is the meatiest method in this module. It takes the Ex. 21 subsidiaries, merges them onto the basic 10k on company name, then takes the overlapping companies and chooses the pair that have the most similar location and report year. In the end, we return the Ex.21 subsidiaries with a CIK merged onto the companies that are determined to file their own 10k.